Hybrid recurrent cache

gabe-l-hart opened the pull request:
This is a re-opened version of #13904 after #13746 was merged

## Description

This PR introduces the `llama_kv_cache_hybrid_recurrent` cache implementation. It follows the pattern of `llama_kv_cache_unified_iswa` by holding two child cache instances and implementing the interface logic such that it manages both correctly for the appropriate layers.

## Changes

The main change in this PR is the addition of `llama_kv_cache_hybrid_recurrent` in `llama-kv-cache-hybrid-recurrent.*`. In addition to this, the PR does the following:

* Add the `llama_model_is_hybrid_recurrent` public API (akin to `llama_model_is_recurrent`)
* Add `LLM_KV_ATTENTION_LAYER_INDICES` as an hparam to hold the indices of the layers that should use attention (versus recurrent)
    * This part is _not_ well aligned with `iswa`, but that mechanism also isn't particularly extensible. It might be more appropriate to have a generic mechanism for indicating the type of caching to use for each layer, but that would start to approach the generic hybrid implementation that I originally attempted which ended up being too abstract (#13276).
* Abstracting utilities in `llm_graph_context` that need a specific type of cache to use getters (`get_state_unified` / `get_state_recurrent`) that will properly handle `llama_kv_cache_hybrid_recurrent`
* Make `n_embd_k_s` / `n_embd_v_s` layer-dependent and use layer indices when calling them in the existing cache implementations
* Add layer filtering to `llama_kv_cache_recurrent`
* Updates the logic in `llama_model::create_memory` to use `llm_arch_is_recurrent` and `llm_arch_is_hybrid_recurrent` rather than relying on adding models to the `switch` statement which was redundant with the implementation of these functions

---

*edits by @ggerganov below:*

Notes so I don't forget

- Naming update for next PRs (https://github.com/ggml-org/llama.cpp/pull/13979#pullrequestreview-2919526914)

gabe-l-hart committed:
feat: Add support for distinguishing recurrent vs non-recurrent layers in hparams

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add c++ side constants for attention layer indices hparam

Branch: GraniteFour

gabe-l-hart committed:
feat: Add llama_model_is_hybrid API call

Also, split llama_model_is_recurrent into llm_arch_is_recurrent in
llama-arch with llama_model_is_recurrent delegating to
llm_arch_is_recurrent. The same split is done for hybird. This is needed
because there are places where the llama_model has not yet been initialized
but we need to check if the model is recurrent (specifically for the
per-layer recurrent check array in hparams).

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Auto-fill hparams.recurrent_layer_arr based on whether the model is recurrent

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Use per-layer sizing everywhere in kv caches

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add layer filter to recurrent cache

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
refactor: rename *_is_hybrid -> *_is_hybrid_recurrent

The implementation of the hybrid cache intentionally does not specify the
types of the child caches, so there was a naming mismatch with these
predicate functions that used "hybrid" to imply "hybrid recurrent."

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Construct hybrid recurrent cache for hybrid recurrent models

This includes a refactor of the create_memory logic to avoid needing to use
the arch enum explicitly unless a model needs explicit cache instantiation
logic beyond the standard logic for recurrent, hybrid, unified, and iswa.

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Fix wrong bool condition for split equal in hybrid cache

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: First pass at llama_kv_cache_hybrid_recurrent

This follows the pattern in iswa where the two child caches are held
explicitly to support the case where a model requires a single attention
cache and a single recurrent cache where each layer uses exactly one of the
caches.

This is a rewrite of the more generic approach in the original hybrid cache
PR: https://github.com/ggml-org/llama.cpp/pull/13276

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Fix shift logic to defer to unified cache

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented on src/llama-graph.cpp:
```diff
@@ -1059,6 +1060,30 @@ ggml_tensor * llm_graph_context::build_pos_bias(ggml_tensor * pos_bucket, ggml_t
     return pos_bias;
 }
 
+const llama_kv_cache_unified_state * llm_graph_context::get_state_unified() const {
+    const auto * umstate = dynamic_cast<const llama_kv_cache_unified_state *>(mstate);
+    if (!umstate) {
+        const auto hmstate = dynamic_cast<const llama_kv_cache_hybrid_recurrent_state *>(mstate);
+        if (hmstate) {
+            umstate = hmstate->get_state_attn();
+        }
+    }
+    GGML_ASSERT(umstate);
+    return umstate;
+}
+
+const llama_kv_cache_recurrent_state * llm_graph_context::get_state_recurrent() const {
+    const auto * rmstate = dynamic_cast<const llama_kv_cache_recurrent_state *>(mstate);
+    if (!rmstate) {
+        const auto hmstate = dynamic_cast<const llama_kv_cache_hybrid_recurrent_state *>(mstate);
+        if (hmstate) {
+            rmstate = hmstate->get_state_recurrent();
+        }
+    }
+    GGML_ASSERT(rmstate);
+    return rmstate;
+}
+
```
I'm working through this now and a couple of questions are coming up:

1. Would it be best to combine `build_inp_s_copy` with `build_attn_inp_kv` for hybrid so that models call just one "build inputs" function, or keep them separate for simplicity?
2. For the `build_attn` methods, each has a corresponding `llm_graph_input_attn_*` class. The `build_inp_s_*` methods don't have this pattern which would make this a bit harder to have code reuse. Are there plans to refactor that further @compilade?
3. In the `mamba2` branch, `s_mask` seems to be totally removed. I'd prefer not to do all of the boilerplate for duplicating `build_inp_s_mask` for the hybrid recurrent case if that's definitely going to be going away. Is there any reason that might stick around past the merge of `mamba2`?

gabe-l-hart replied:
Thanks, that's super helpful! I was missing the distinction between `build_attn_inp` and `build_attn` which makes perfect sense.

>Personally, I think it would be simpler to keep them separate

I agree on my personal gut feeling, so I'll go with this and see how it feels once complete.

gabe-l-hart replied:
Ok, I think this feels a lot cleaner now. For `build_inp_s_copy`, I opted to add an optional parameter so that the caller can take ownership of casting the cache state rather than duplicating the function into `build_inp_s_copy_hybrid`. That felt a little cleaner w.r.t. code reuse, but I'm happy to do a separate method if that's preferred.

gabe-l-hart committed:
feat: Support hybrid recurrent in llama-graph

NOTE: I intentionally did not add support for s_mask since it will be going
away soon

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Fix logic for initializing inputs and attn layers for hybrid caches

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Update recurrent cache for changes to remove intermediate kv_cache interface

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Fix status for init_update sig for recurrent cache state

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Add missing padding to n_ctx for hybrid cache construction

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Update clear signature for data argument after rebase

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
@ggerganov I've noticed that the `Server` tests are [consistently failing](https://github.com/ggml-org/llama.cpp/actions/runs/15539011821/job/43744995814?pr=13979) on this branch, but I see them passing on other PRs (eg https://github.com/ggml-org/llama.cpp/pull/14081). The failures seem to be around retried connections to `/slots/0?action=restore` ([here](https://github.com/ggml-org/llama.cpp/actions/runs/15539011821/job/43744995814?pr=13979#step:13:159)). I've noticed a couple of things when trying to repro locally:

1. For all of the [server tests](https://github.com/ggml-org/llama.cpp/tree/master/tools/server/tests), despite the [health check ping loop](https://github.com/ggml-org/llama.cpp/blob/master/tools/server/tests/utils.py#L228), I need to manually inject a `sleep` after `server.start()` to get any of the tests to pass on my M3 Max 64GB, otherwise the first endpoint call will almost always return an unexpected result (invalid json, or 5xx error). To me, this speaks to the server reporting `/health` success before it's logically "ready." This is probably correct in that the server is healthy enough to take requests, but the tests are treating this as a readiness check which would indicate that the server is fully initialized which seems incorrect (similar to [liveness vs readiness in kubernetes](https://medium.com/@sachinadi424/keeping-kubernetes-healthy-liveness-and-readiness-probe-explained-661757b825ff)). I wasn't able to find any open issues related to this, so I'd be curious if there's a discussion somewhere of readiness vs liveness?

2. Once I _do_ insert the manual `sleep` in `test_slot_save_restore`, the test passes locally for me. This speaks to some difference between my local system and the GH Action runners. I'm not familiar enough with the server code to know whether this would be caused by some kind of memory constraint, disk speed constraint, etc, but my guess would be something like `slot1.bin` not being fully written somehow. This wouldn't explain why it passes on other branches though. Any insight on the nature of these tests would be much appreciated as I try to debug!

gabe-l-hart commented:
I've tried adding retry logic for all requests in https://github.com/ggml-org/llama.cpp/pull/13979/commits/39a93b3915c678a9047a3a1b85cae66e8437b8ec to work around the race between `/health`, but I'm not sure if this is just going to mask an underling issue.

ggerganov commented:
The changes to the server tests should not be needed. Let's revert the commit for now and I'll investigate.

gabe-l-hart commented:
Issue for follow up on `/health` race condition with tests: https://github.com/ggml-org/llama.cpp/issues/14092

gabe-l-hart commented:
That was quick! Undrafting now that #13834 is merged

gabe-l-hart commented on src/llama-graph.h:
```diff
@@ -242,7 +243,7 @@ class llm_graph_input_attn_kv_unified : public llm_graph_input_i {
         kv_state(kv_state) {
     }
-    ~llm_graph_input_attn_kv_unified() = default;
+    virtual ~llm_graph_input_attn_kv_unified() = default;
```
I don't think this `virtual` is needed. I think I added it when I was attempting to have the hybrid input inherit from the unified input

gabe-l-hart committed:
fix: Remove errant virtual destructor leftover from previous impl attempt

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Use per-layer n_embd_k/v_s calls for mamba (1) layers

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented on src/llama-hparams.h:
```diff
@@ -181,10 +184,13 @@ struct llama_hparams {
     // dimension of the rolling state embeddings
     // corresponds to Mamba's conv_states size or RWKV's token_shift states size
-    uint32_t n_embd_k_s() const;
+    uint32_t n_embd_k_s(uint32_t il = 0) const;
```
I initially added these with the default value of `0` to avoid needing to change code that had no chance of being in the path of a hybrid model. After this PR, I think the only place that's true is the various `rwkv` methods in `llama-graph` and `llama-model`. Would it be better to always require an argument?

gabe-l-hart commented on src/llama-graph.h:
```diff
@@ -594,12 +597,13 @@ struct llm_graph_context {
+                                     int32_t   state_size,
+                                     int32_t   n_seqs,
+                                        bool   avoid_copies = false,
+        const llama_kv_cache_recurrent_state * kv_state = nullptr) const;
```
NIT: This was the closest I could come up to for consistent indentation, but it's... ugly? Very open to preferred style suggestions.

Non-NIT: I don't love adding this, especially after `avoid_copies` since it now forces all layer builders that may be in a hybrid architecture to explicitly set `avoid_copies`.

gabe-l-hart commented on src/llama-kv-cache-hybrid-recurrent.cpp:
```diff
@@ -0,0 +1,241 @@
+
+    // prepare the recurrent batches first
+    if (!kv_recurrent->prepare(ubatches)) {
+        // TODO: will the recurrent cache be in an undefined state at this point?
+        LLAMA_LOG_ERROR("%s: failed to prepare recurrent ubatches\n", __func__);
```
Should I strip out this TODO at this point?

gabe-l-hart commented on src/llama-kv-cache-hybrid-recurrent.cpp:
```diff
@@ -0,0 +1,245 @@
+}
+
+llama_pos llama_kv_cache_hybrid_recurrent::seq_pos_min(llama_seq_id seq_id) const {
+    // the min of the total cache is the max of the two caches' min values
```
I'd definitely love a double-check on this logic!

compilade replied:
I think this is fine, since for the recurrent cache, `seq_pos_min` and `seq_pos_max` are equal in practice for now, because there's only one cell per sequence instead of one cell per token (as in the unified self-attention KV cache).

It might not really make sense to have a "minimal sequence position" for the recurrent cache.

Using the largest minimum will remain good if/when `llama_kv_cache_recurrent::seq_pos_min` is changed to return 0 for existing `seq_ids` instead of a position in the cells. (since conceptually the state does contain some information from the beginning even without an extra state checkpoint (multiple recurrent cells per sequences were implemented at some point in #7531, but not on `master`))

gabe-l-hart commented on src/llama-kv-cache-hybrid-recurrent.h:
```diff
@@ -0,0 +1,144 @@
+    // llama_kv_cache_hybrid_recurrent specific API
+    //
+
+    llama_kv_cache_unified   * get_kv_attn     () const;
```
I don't think I've ever actually used the `friend` keyword in real code, but I think we _could_ make `llama_kv_cache_hybrid_recurrent_state` a `friend` and then make these `private` for slightly better encapsulation.

gabe-l-hart replied:
Actually, if we made it a `friend`, these methods could just be removed since the child `unique_ptr` instances could just be accessed directly.

gabe-l-hart commented on src/llama-kv-cache-recurrent.cpp:
```diff
@@ -88,8 +94,8 @@ llama_kv_cache_recurrent::llama_kv_cache_recurrent(
         ggml_tensor * v = ggml_new_tensor_1d(ctx, type_v, n_embd_v_gqa*kv_size);
         ggml_format_name(k, "cache_k_l%d", i);
         ggml_format_name(v, "cache_v_l%d", i);
-        k_l.push_back(k);
```
This `push_back` was broken with the addition of the filter above to skip non-recurrent layers.

gabe-l-hart commented on src/llama-kv-cache-recurrent.h:
```diff
@@ -15,13 +15,18 @@
 class llama_kv_cache_recurrent : public llama_memory_i {
 public:
+
+    // this callback is used to filter out layers that should not be included in the cache
```
This is copy-pasta from the `unified` cache. I had originally hoisted it to the `llama_kv_cache` intermediate base class, but since that's been removed, I figured it was better to just copy it rather than put it into the top-level interface class.

gabe-l-hart commented on src/llama-model.cpp:
```diff
@@ -469,6 +470,10 @@ void llama_model::load_hparams(llama_model_loader & ml) {
     std::fill(hparams.n_head_arr.begin(),    hparams.n_head_arr.end(),    0);
     std::fill(hparams.n_head_kv_arr.begin(), hparams.n_head_kv_arr.end(), 0);
     std::fill(hparams.n_ff_arr.begin(),      hparams.n_ff_arr.end(),      0);
+    std::fill(
```
On this branch, this is somewhat incomplete. For the Granite 4 branch, I have this array further overridden by parsing `LLM_KV_ATTENTION_LAYER_INDICES`. I'm wondering if that logic should be universal enough that I should hoist it to this portion here to be done in a model-agnostic way.

gabe-l-hart commented on src/llama-model.cpp:
```diff
@@ -13198,6 +13203,8 @@ llama_memory_i * llama_model::create_memory(const llama_memory_params & params,
     llama_memory_i * res;
 
     switch (arch) {
+        // Models that need specific instantiation should be handled in the
```
This is a slight architectural refactor to avoid the need to duplicate the arch switch statement between this and `llm_arch_is_[hybrid]_recurrent`. I'm not strictly tied to this, but it feels a little cleaner to hold a single source-of-truth for determining which models are hybrid/hybrid-recurrent.

gabe-l-hart commented:
@compilade @ggerganov I think this one is ready for a full review now based on [the list](https://github.com/ggml-org/llama.cpp/pull/13746#issuecomment-2921703899) shared on the earlier refactor PR

younesbelkada commented on src/llama-kv-cache-hybrid-recurrent.cpp:
```diff
@@ -0,0 +1,245 @@
+             uint32_t   n_seq_max,
+                 bool   offload) :
+    hparams(model.hparams),
+    kv_attn(new llama_kv_cache_unified(
```
Hi @gabe-l-hart - thanks for your huge work. Chiming in this PR to ask a question: how do you think we should handle cache formats where for every layer we have both the attention and SSM states ? This design seems to be tailored for architectures that have either SSM or Attention for each decoder block

younesbelkada replied:
Thank you again for all your huge work, we have been following all these PRs recently to see how we can adapt it for H1 and it was very helpful. Happy to work on addressing the comments you mentioned or anything that can help adding these new architectures in the future 🙏

gabe-l-hart replied:
Awesome, I'm reading up on Falcon H1 and will see if there are any tweaks that make sense to incorporate now to avoid refactors later. I'll let you know how far I get!

ggerganov replied:
> Hi @gabe-l-hart - thanks for your huge work. Chiming in this PR to ask a question: how do you think we should handle cache formats where for every layer we have both the attention and SSM states ? This design seems to be tailored for architectures that have either SSM or Attention for each decoder block

@younesbelkada The explanation by @gabe-l-hart sounds good. Regarding this:

> Creating another new subclass of llama_memory_i that owns the child caches (similar to hybrid_recurrent)
The various interfaces would probably be implemented with "AND" semantics instead of "OR" semantics for child layers

I think instead of introducing a new class, we can simply extend the hybrid implementation to support the H1 architecture. It should be just a matter of passing the correct layer filters.

> The place where I see the most incompatibility is in the hparams.recurrent_layer() method which right now is a "yes/no" and if true implies ! attention_layer (explicitly in the n_embd_k_s / n_embd_v_s methods).

Yes, probably we have to introduce the notion of `has_recurrent_state(il)`, `has_attention(il)` and these can be both `true` in such cases.

> In terms of implementation in llama.cpp, I think the only things that would need to change would be how the hparams interact with the cache.

Yup, sounds about right.

gabe-l-hart replied:
>Yes, probably we have to introduce the notion of has_recurrent_state(il), has_attention(il) and these can be both true in such cases.

This makes sense. I'll give this a shot soon.

>Of these uses, I think the two that would be the most problematic are n_embd_k/v_s since they're currently [implemented using hparams.recurrent_layer](https://github.com/gabe-l-hart/llama.cpp/blob/HybridRecurrentCache/src/llama-hparams.cpp#L69) and hparams.n_embd since that is used by both the unified and recurrent caches and I think might need to have different values for the two if they were being shared by the same layer.

@ggerganov with the removal of `n_embd_k/v_s` layer indexing, I think that just leaves `hparams.n_embd`. Do you have a quick thought on whether this will need to be split between attn/recurrent portions for a single layer?

gabe-l-hart replied:
I'm not clear on that (haven't unboxed it far enough either). @younesbelkada I haven't dug into your fork deeply yet, but I'm guessing you may have had to tackle this question and might know which hparams would need to be different between the attention / recurrent portions of the cache for a given layer.

gabe-l-hart replied:
FYI @younesbelkada based on the big refactor in the `llm_graph_context` interfaces requested in the other thread, I've just redone `manba2-sync` and force pushed, so if you're working off that branch, you may have a little refactoring to do. The main change is to move from `build_inp_s_copy` -> `build_rs_inp_recurrent` / `build_rs_inp_hybrid_recurrent` and then `build_recurrent_state` -> `build_rs`.

younesbelkada replied:
Thank you @gabe-l-hart - I had a look by working from your latest changes of that branch. Somehow I am facing few issues and I thought you might have some ideas on what went wrong. 
in `llm_build_falcon_h1` I initialize two inputs:

```
    llm_build_falcon_h1(const llama_model & model, const llm_graph_params & params, ggml_cgraph * gf) : llm_graph_context(params), model(model)  {
        const int64_t n_embd_head = hparams.n_embd_head_v;

        ggml_tensor * cur;
        ggml_tensor * inpL;

        inpL = build_inp_embd(model.tok_embd);
        inpL = ggml_scale(ctx0, inpL, hparams.embedding_multiplier);

        // inp_pos - contains the positions
        ggml_tensor * inp_pos = build_inp_pos();
        
        auto * inp_attn_hybrid = build_attn_inp_kv_hybrid_recurrent();
        auto * inp_rs_hybrid = build_rs_inp_hybrid_recurrent();
```
Somehow if I comment `auto * inp_rs_hybrid = build_rs_inp_hybrid_recurrent();` and keep only the attention layer, inference works smoothly (while getting giberish of course), however, when I comment out the line, I face segmentation errors, even if `inp_rs_hybrid` is not used at all. 
Another point is that I had to replace:
```
    k_l.reserve(n_layer);
    v_l.reserve(n_layer);
```
with
```
    k_l.resize(n_layer);
    v_l.resize(n_layer);
``` 
Inside `llama-kv-cache-recurrent.cpp` for some reason, otherwise 0.0MiB memory will be allocated for the SSM cache. I would like to have your review to see if my approaches are correct. Would it be ok if I open a draft PR between your mamba2 synced branch and mine to help me figure out if the changes look correct and if I missed anything ? Thank you very much !

gabe-l-hart replied:
Very good catch on `resize` vs `reserve`. This is what I get for wading back into `c++` and thinking I still remember the `stl` after many years away 🤦.

>when I comment out the line, I face segmentation errors, even if inp_rs_hybrid is not used at all.

Hm, this sounds really plausible. I've seen this kind of issue before and always ended up just firing it up with `lldb` (I'm on a Mac for dev work, so `gdb` for linux, no idea for Windows). Poking through with the debugger usually reveals the issue for me.

>Would it be ok if I open a draft PR between your mamba2 synced branch and mine to help me figure out if the changes look correct and if I missed anything ?

Absolutely! It will likely be a moving target since I've been trying to continuously rebase as `master` keps moving, so it's very possible your PR will be broken frequently until the upstream parts get merged, but please feel free to do whatever works best for you to keep track of all the moving branches.

younesbelkada replied:
Thank you very much @gabe-l-hart - I was able to finally get inference at least running without errors, your changes were really helpful ! Will open a draft PR soon

gabe-l-hart committed:
refactor: Remove n_embd_k/v_s from unified cache

No longer needed now that unified isn't also supporting recurrent

https://github.com/ggml-org/llama.cpp/pull/13979#discussion_r2140761069

Branch: HybridRecurrentCache

gabe-l-hart committed:
refactor: Remove layer index from n_embd_k/v_s

Now that it's not used at all in the unified cache, we don't need to use
the layer index to zero it out for attention layers.

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented on src/llama-kv-cache-recurrent.cpp:
```diff
@@ -69,8 +69,8 @@ llama_kv_cache_recurrent::llama_kv_cache_recurrent(
-        const uint32_t n_embd_k_gqa = hparams.n_embd_k_gqa(i) + hparams.n_embd_k_s(i);
-        const uint32_t n_embd_v_gqa = hparams.n_embd_v_gqa(i) + hparams.n_embd_v_s(i);
+        const uint32_t n_embd_k_gqa = hparams.n_embd_k_gqa(i) + hparams.n_embd_k_s();
+        const uint32_t n_embd_v_gqa = hparams.n_embd_v_gqa(i) + hparams.n_embd_v_s();
```
I _think_ that sounds right. It looks like the value here is ultimately determined by `hparams.n_head_kv_arr` which by default is populated as a copy of `hparams.n_head_arr` from `%s.attention.head_count`. For `mamba`, I see `mamba.attention.head_count` set to `0` and I suspect this would be true of all other recurrent models, but I'm not sure how to fully verify. I'll try removing them and see if the recurrent models I have work as expected, but I'll defer to @compilade on whether this is an "always" thing or a "just mamba* models" thing.

gabe-l-hart committed:
refactor: Remove n_embd_k/v_gqa from recurrent cache

This is no longer needed now that there are separate implementations

https://github.com/ggml-org/llama.cpp/pull/13979#discussion_r2140825128

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Allow custom layer filters for hybrid recurrent

This should help support architectures like Falcon H1 where there is
overlap between layers that need attention and recurrent caches.

https://github.com/ggml-org/llama.cpp/pull/13979#discussion_r2140748922

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov reviewed (commented):
I feel like the naming of the recurrent cache is not great and it is causing some confusion. In a follow-up PR we have to do the following naming changes

- Rename `llama_kv_cache_recurrent` -> `llama_recurrent_state`
- Rename `llama_kv_cache_hybrid_recurrent` -> `llama_memory_hybrid`

This should be a good first step, although not ideal because "hybrid" will implicitly refer to "unified KV cache + recurrent state". For the future we have to figure out a naming scheme that would be able to differentiate between different types of hybrid memories. But this should be good enough for now.

gabe-l-hart committed:
fix: Remove logits_all after rebase

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Remove llama_model_is_hybrid_Recurrent public API

https://github.com/ggml-org/llama.cpp/pull/13979#discussion_r2141728423

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
refactor: Use llama_memory_state_ptr for child states in hybrid memory state

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented on src/llama-graph.h:
```diff
@@ -508,7 +509,7 @@ struct llm_graph_context {
     ggml_tensor * build_inp_mean() const;
     ggml_tensor * build_inp_cls() const;
-    ggml_tensor * build_inp_s_copy() const;
+    ggml_tensor * build_inp_s_copy(const llama_kv_cache_recurrent_state * kv_state = nullptr) const;
```
Ok, this makes sense. I'll work on this refactor which I think will also fix the comment below about needing to pull the `kv_state` from the `inp` class.

gabe-l-hart replied:
I'm trying to work through all of the changes suggested here and the implications above of not mixing `unified` with `hybrid` in `build_attn`. It's resulting in a lot of code duplication, most of which seems fairly benign, but I'm now to the point of implementing `build_rs(llm_graph_input_rs_hybrid * inp, ...` which will be identical to the non-hybrid version except for the `kv_state` extraction. As is today, it's a pretty complex method, so I worry that if issues arise in the logic, they will now need to be fixed twice.

One alternative that _might_ work and would reduce a lot of the changes in `llama_graph` would be for `llama_kv_cache_hybrid_recurrent` state to use multiple inheritance and use an `isa` approach to the child state types rather than a `hasa` approach. This would require that the child state types declare their type-specific interface methods as `virtual`, but it would avoid the need to differentiate state types everywhere in `llama-graph`.

For completeness, I'm going to try to push forward with the current proposal of creating `hybrid` versions of all structs and methods in `llama-graph`, but I wanted to throw this out as an option to consider to reduce the maintenance burden.

gabe-l-hart replied:
Another option that would flagrantly violate the [coding guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md#coding-guidelines) would be to use template specialization for the various implementations of `build_attn` / `build_rs`. This would allow a common implementation where appropriate that delegates the "get the state" functionality to an implementation-detail function that is overloaded by type.

(not suggesting it, just brainstorming)

gabe-l-hart replied:
Ok, I've now got the two sets of classes / methods to be quite consistent with the duplicate `build_attn` / `build_rs` implementations to avoid abstracting over the memory type

gabe-l-hart replied:
I'm trying to rebase Granite 4 now and it's pushing the code duplication further to require that the layer builder methods (`build_mamba_layer`, `build_mamba2_layer`) be duplicated so that they can take the right `inp` type.

I think there might be a _much_ simpler solution though: Override the type cast operator for the hybrid state so that it correctly type casts to its children with `static_cast`. I think this would completely remove the need for `hybrid_recurrent` versions of any of the graph methods!

I need to sign off for the day, but will try this tomorrow if there are no objections.

gabe-l-hart replied:
Ok, I'm pushing on this now, but I keep coming back to the `static_cast<>` baked into the `build_rs` methods. Even if I write a shared `build_rs_common` (name TBD) method that simply takes the right dereferenced values off of the correctly cast `kv_state`, we still need multiple `build_rs` implementations that dispatch based on the type of the `inp` argument (following the pattern). This in turn pushes up to the `build_mamba*_layer` methods causing the same problem over again. The `build_rs` calls in those methods are buried inside hundreds of other lines of code, so unlike how `build_attn_mha` is the "heart" of the attention mechanisms and the surrounding code is relatively trivial, for the recurrent layer builders, the `build_rs` is only a small part of the complexity. Duplicating all of that surrounding code based on the `inp` type would be a lot of complexity to maintain. We could simply template those methods on the `inp` type, but that would violate the desire to avoid templates.

As an alternative while we think through these refactors, I'm _fairly_ certain that if the `llama_kv_cache_hybrid_recurrent_state` simply provided cast operators to the two child state types, all changes in `llama-graph*` could be reverted from this PR and we could tackle this refactoring separately.

ggerganov replied:
Looking at the `build_mamba_layer()` function, it appears complex, but actually it boils down to something like this:

```c++
build_mamba_layer(inp) {
    rec_state = static_cast<...>(mstate);

    conv = rec_state->update_conv(inp); // analogous to unified get_k() + cpy_k()
    ssm  = rec_state->update_ssm(inp);  // analogous to unified get_v() + cpy_v()

    common = get_common(model);

    do_something_with_conv(model, common, conv);
    do_something_with_ssm (model, common, ssm);
}
```

The `get_common(model)` and `do_something_with_...(model, ...)` parts have to be extracted out of this function and remain implemented in `llama-model.cpp` because they use model tensors and do not depend on the memory module. The `update_conv()` and `update_ssm()` function have to be implemented only once in the `class llama_kv_cache_recurrent`. Internally, they can use a helper `build_recurrent_state()` function to deduplicate the code. Note that the hybrid cache will call these methods from the recurrent cache instance and won't have to implement them a second time.

When you decompose the function like this, the implementation becomes much simpler:

```c++
build_mamba_layer(inp) {
    rec_state = static_cast<...based_on_inp...>(m_state);

    conv = rec_state->update_conv();
    ssm  = rec_state->update_ssm();

    return { conv, ssm };
}
```

This is much easier to overload multiple times for different `inp` types. I would even split it one more time:

```c++
// llama-graph:

// one for recurrent and one for hybrid
build_mamba_conv(inp) {
    rec_state = static_cast<...based on inp...>(m_state);

    conv = rec_state->update_conv();
    
    return conv;
}

// one for recurrent and one for hybrid
build_mamba_ssm(inp) {
    rec_state = static_cast<...based on inp...>(m_state);

    ssm = rec_state->update_ssm();
    
    return ssm;
}
```

I hope I'm not missing something, but I think this should work and should be clean.

Alternatively, we can also just bring the PR to a working state any way you think makes sense and then I will try to do a refactoring pass and see how it goes. Would just need some sample test commands to experiment with.

compilade replied:
> The `update_conv()` and `update_ssm()` function have to be implemented only once in the `class llama_kv_cache_recurrent`.

These would be very specific to Mamba, and I don't think model-specific graph-building functions belong in `llama_kv_cache_recurrent`.

As described (`analogous to unified get_v() + cpy_v()`), `update_ssm()` would need tensors from `model.layers`, unless split into more functions (`update_conv()` doesn't have this problem).

From my understanding, `update_ssm()` would fetch, modify, and write back the states for the ssm, but the modifying step involves `ggml_ssm_scan`. I might have misunderstood.

> I hope I'm not missing something, but I think this should work and should be clean.

The `update_ssm()` and `do_something_with_ssm()` would be very much intertwined, especially since <https://github.com/ggml-org/llama.cpp/commit/2fa5f2ceb8b49bbd2835878ad5429ea74383566c> (in #9126) replaces `avoid_copies` with a passed function in `build_recurrent_state` to make it work properly:

https://github.com/ggml-org/llama.cpp/blob/0b6f6becb4e916a24fcaf2966647381a21d1f084/src/llama-model.cpp#L9220-L9231

which fixes the `FIXME` added in <https://github.com/ggml-org/llama.cpp/pull/13834/commits/62a9f34baefc657212dea8f1bad14d4fb1657da8> (see also <https://github.com/ggml-org/llama.cpp/pull/13834#issuecomment-2957652278>)

https://github.com/ggml-org/llama.cpp/blob/62a9f34baefc657212dea8f1bad14d4fb1657da8/src/llama-graph.cpp#L1447-L1457

---

What bothers me is that fundamentally, an hybrid state doesn't/shouldn't require re-implementing graph-building functions from the inner state types, since those inner state types are independent in principle.

> Hmm, I may be wrong about the cast operator. I'll keep digging.

I'm likely wrong, but this is possibly caused by how `mstate` in `llm_graph_context` is a `llama_memory_state_i` which doesn't have virtual type cast operators which means the type-cast operators of the child classes won't be used except maybe with a `dynamic_cast` (but that may be slow (or not?)).

gabe-l-hart replied:
Thanks for all the thoughts / ideas! At the moment, I'm trying to do the "simple with lots of duplication" version and I have it back to the [broken state](https://github.com/ggml-org/llama.cpp/pull/9126#issuecomment-2963750982) with Granite 4 that it's been in since the latest changes on the `mamba2` branch. I've been digging on that issue, so will try to resolve what I'm doing wrong there, then see about how much I can simplify the duplicate code on this branch (and the corresponding duplicate code on Granite 4).

>What bothers me is that fundamentally, an hybrid state doesn't/shouldn't require re-implementing graph-building functions from the inner state types, since those inner state types are independent in principle.

I think this is where my head is too. If the hybrid cache could simply _be_ both child types, it could "just work" everywhere the children do. I had originally tried to implement this with multiple (diamond) inheritance, but was running into the same issues I'm seeing with the same thing in the models where the common parent ends up getting double-initialized (I feel like I'm missing something important around `virtual` inheritance).

>I'm likely wrong, but this is possibly caused by how mstate in llm_graph_context is a llama_memory_state_i which doesn't have virtual type cast operators which means the type-cast operators of the child classes won't be used except maybe with a dynamic_cast (but that may be slow (or not?)).

I think this likely is the issue. I've been playing with a dummy version to try to understand what's going on:

```c++
// Example program
#include <iostream>
#include <string>

struct Base {
    virtual void foo() const = 0;
};

struct A : public Base {
    int i;

    A(int i) : i(i) {}

    void foo() const override {
        std::cout << "A::foo(" << i << ")" << std::endl;
    }
};

struct B : public Base {
    std::string s;
    
    B(std::string s) : s(s) {}

    void foo() const override {
        std::cout << "B::foo(" << s.c_str() << ")" << std::endl;
    }
};

struct C : public Base {
    C(bool be_a, int i, std::string s) : be_a(be_a), a(i), b(s) {}

    void foo() const override {
        std::cout << "C::foo" << std::endl;
        if (be_a) a.foo();
        else b.foo();
    }

    explicit operator const A*() const { return &a; }
    explicit operator const B*() const { return &b; }
private:
    const bool be_a;
    const A a;
    const B b;
};

void use_a_as_base(const Base * b) {
    // Doesn't use overloaded operators because the base
    // doesn't have them as virtual
    const A * a = static_cast<const A *>(b);
    std::cout << "Using a from Base: " << a->i << std::endl;
}

// void use_a(const C * c) {
//     // Doesn't compile because C doesn't inherit from A
//     const A * a = static_cast<const A *>(c);
//     std::cout << "Using a: " << a->i << std::endl;
// }

int main()
{
    C ca(true, 1, "foo");
    C cb(false, 2, "bar");
    ca.foo();
    cb.foo();
    
    // Bad access of cb's non-existent `i`
    use_a_as_base(&cb);
}
```

```
C::foo
A::foo(1)
C::foo
B::foo(bar)
Using a from Base: 0
```

([link to sandbox](https://cpp.sh/?source=%2F%2F+Example+program%0A%23include+%3Ciostream%3E%0A%23include+%3Cstring%3E%0A%0Astruct+Base+%7B%0A++++virtual+void+foo()+const+%3D+0%3B%0A%7D%3B%0A%0Astruct+A+%3A+public+Base+%7B%0A++++int+i%3B%0A%0A++++A(int+i)+%3A+i(i)+%7B%7D%0A%0A++++void+foo()+const+override+%7B%0A++++++++std%3A%3Acout+%3C%3C+%22A%3A%3Afoo(%22+%3C%3C+i+%3C%3C+%22)%22+%3C%3C+std%3A%3Aendl%3B%0A++++%7D%0A%7D%3B%0A%0Astruct+B+%3A+public+Base+%7B%0A++++std%3A%3Astring+s%3B%0A++++%0A++++B(std%3A%3Astring+s)+%3A+s(s)+%7B%7D%0A%0A++++void+foo()+const+override+%7B%0A++++++++std%3A%3Acout+%3C%3C+%22B%3A%3Afoo(%22+%3C%3C+s.c_str()+%3C%3C+%22)%22+%3C%3C+std%3A%3Aendl%3B%0A++++%7D%0A%7D%3B%0A%0Astruct+C+%3A+public+Base+%7B%0A++++C(bool+be_a%2C+int+i%2C+std%3A%3Astring+s)+%3A+be_a(be_a)%2C+a(i)%2C+b(s)+%7B%7D%0A%0A++++void+foo()+const+override+%7B%0A++++++++std%3A%3Acout+%3C%3C+%22C%3A%3Afoo%22+%3C%3C+std%3A%3Aendl%3B%0A++++++++if+(be_a)+a.foo()%3B%0A++++++++else+b.foo()%3B%0A++++%7D%0A%0A++++explicit+operator+const+A*()+const+%7B+return+%26a%3B+%7D%0A++++explicit+operator+const+B*()+const+%7B+return+%26b%3B+%7D%0Aprivate%3A%0A++++const+bool+be_a%3B%0A++++const+A+a%3B%0A++++const+B+b%3B%0A%7D%3B%0A%0Avoid+use_a_as_base(const+Base+*+b)+%7B%0A++++%2F%2F+Doesn%27t+use+overloaded+operators+because+the+base%0A++++%2F%2F+doesn%27t+have+them+as+virtual%0A++++const+A+*+a+%3D+static_cast%3Cconst+A+*%3E(b)%3B%0A++++std%3A%3Acout+%3C%3C+%22Using+a+from+Base%3A+%22+%3C%3C+a-%3Ei+%3C%3C+std%3A%3Aendl%3B%0A%7D%0A%0A%2F%2F+void+use_a(const+C+*+c)+%7B%0A%2F%2F+++++%2F%2F+Doesn%27t+compile+because+C+doesn%27t+inherit+from+A%0A%2F%2F+++++const+A+*+a+%3D+static_cast%3Cconst+A+*%3E(c)%3B%0A%2F%2F+++++std%3A%3Acout+%3C%3C+%22Using+a%3A+%22+%3C%3C+a-%3Ei+%3C%3C+std%3A%3Aendl%3B%0A%2F%2F+%7D%0A%0Aint+main()%0A%7B%0A++++C+ca(true%2C+1%2C+%22foo%22)%3B%0A++++C+cb(false%2C+2%2C+%22bar%22)%3B%0A++++ca.foo()%3B%0A++++cb.foo()%3B%0A++++%0A++++%2F%2F+Bad+access+of+cb%27s+non-existent+%60i%60%0A++++use_a_as_base(%26cb)%3B%0A%7D))

gabe-l-hart committed:
feat: Overhaul build_recurrent_state / build_inp_s_copy to match attention pattern

https://github.com/ggml-org/llama.cpp/pull/13979/files#r2141701738

This is a big overhaul to bring consistency between how inputs and per-
layer components are created for attention layers and recurrent layers. The
main changes are:

- Rename class llm_graph_input_s_copy -> llm_graph_input_rs
- Add a corresponding llm_graph_input_rs_hybrid_recurrent
- Rename build_inp_s_copy -> build_rs_inp_recurrent
- Add a corresponding build_rs_inp_hybrid_recurrent
- Rename build_recurrent_state -> build_rs to match build_attn w/
llm_graph_input_rs android-build AUTHORS bamba-9b-2.2T.gguf bamba-9b-2.2T.q4_k_m.gguf broken.log build build-rel build-xcframework.sh build.android build.android.bak ci cmake CMakeLists.txt CMakePresets.json CODEOWNERS common common.o CONTRIBUTING.md convert_hf_to_gguf_update.py convert_hf_to_gguf.py convert_llama_ggml_to_gguf.py convert_lora_to_gguf.py debug.log docs examples flake.lock flake.nix ggml ggml-alloc.o ggml-backend.o ggml-metal.o ggml-model-BF16.gguf ggml-model-Q4_K_M.gguf ggml-quants.o ggml.o gguf-py grammar-parser.o grammars include LICENSE licenses llama.log llama.o llamacpp_trace.log main.log Makefile media models mypy.ini pocs poetry.lock prompts pyproject.toml pyrightconfig.json q4_k_m_boot.log q8_0_boot.log quant.log quant2.log README.md requirements requirements.txt sampling.o scripts SECURITY.md src test-grammar-output.tmp test-json-schema-input.tmp tests tools vendor working.log as the first input
- Add a corresponding overload of build_rs w/
llm_graph_input_rs_hybrid_recurrent android-build AUTHORS bamba-9b-2.2T.gguf bamba-9b-2.2T.q4_k_m.gguf broken.log build build-rel build-xcframework.sh build.android build.android.bak ci cmake CMakeLists.txt CMakePresets.json CODEOWNERS common common.o CONTRIBUTING.md convert_hf_to_gguf_update.py convert_hf_to_gguf.py convert_llama_ggml_to_gguf.py convert_lora_to_gguf.py debug.log docs examples flake.lock flake.nix ggml ggml-alloc.o ggml-backend.o ggml-metal.o ggml-model-BF16.gguf ggml-model-Q4_K_M.gguf ggml-quants.o ggml.o gguf-py grammar-parser.o grammars include LICENSE licenses llama.log llama.o llamacpp_trace.log main.log Makefile media models mypy.ini pocs poetry.lock prompts pyproject.toml pyrightconfig.json q4_k_m_boot.log q8_0_boot.log quant.log quant2.log README.md requirements requirements.txt sampling.o scripts SECURITY.md src test-grammar-output.tmp test-json-schema-input.tmp tests tools vendor working.log as the first input
- Add a llm_graph_input_attn_kv_hybrid_recurrent analogous to
llm_graph_input_attn_kv_unified
- Add a build_attn override that takes
llm_graph_input_attn_kv_hybrid_recurrent android-build AUTHORS bamba-9b-2.2T.gguf bamba-9b-2.2T.q4_k_m.gguf broken.log build build-rel build-xcframework.sh build.android build.android.bak ci cmake CMakeLists.txt CMakePresets.json CODEOWNERS common common.o CONTRIBUTING.md convert_hf_to_gguf_update.py convert_hf_to_gguf.py convert_llama_ggml_to_gguf.py convert_lora_to_gguf.py debug.log docs examples flake.lock flake.nix ggml ggml-alloc.o ggml-backend.o ggml-metal.o ggml-model-BF16.gguf ggml-model-Q4_K_M.gguf ggml-quants.o ggml.o gguf-py grammar-parser.o grammars include LICENSE licenses llama.log llama.o llamacpp_trace.log main.log Makefile media models mypy.ini pocs poetry.lock prompts pyproject.toml pyrightconfig.json q4_k_m_boot.log q8_0_boot.log quant.log quant2.log README.md requirements requirements.txt sampling.o scripts SECURITY.md src test-grammar-output.tmp test-json-schema-input.tmp tests tools vendor working.log as the first input

This makes the two paradigms fully consistent. The main drawback is the
code duplication in the build_attn and build_rs implementations where the
only difference between implementations is how they cast the memory state.

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented on src/llama-kv-cache-hybrid-recurrent.cpp:
```diff
@@ -0,0 +1,252 @@
+        n_seq_max
+    )) {}
+
+llama_memory_state_ptr llama_kv_cache_hybrid_recurrent::init_batch(const llama_batch & batch, uint32_t n_ubatch, bool embd_pooled) {
```
No idea if this is relevant, but if I'm reading [this line](https://github.com/ggml-org/llama.cpp/pull/14130/files#diff-d237aaf22d44f95e255449338aee870e6fc21c85b002789962fc7b3c30c5fe15L833) correctly, I think the index into `data` changed in #14130 because you end up multiplying `s*n_seq_tokens` by `n_kv` whereas it wasn't multiplied before the change.

gabe-l-hart replied:
I also see https://github.com/ggml-org/llama.cpp/pull/14119 touching the same areas, so it's also possibly related

gabe-l-hart replied:
I've debugged the bad output to [this line](https://github.com/gabe-l-hart/llama.cpp/blob/HybridRecurrentCache/src/llama-graph.cpp#L1127). The previous op ([here](https://github.com/gabe-l-hart/llama.cpp/blob/HybridRecurrentCache/src/llama-graph.cpp#L1100)) has matching inputs before/after the breakage, but the output of the `ggml_soft_max_ext` op is different.

I've stepped through the op in cpu mode and I'm pretty confident I've verified that the tensor data in `kq_mask` is different. In the not-broken run, the only cells with `0` instead of `-Inf` are in the first position of each row, whereas in the broken run, there are `0`s in other positions (eg both `4096` and `4097` whereas `4097` is `-Inf` in the not-broken run).

This leaves me pretty sure that something in how the `kq_mask` is initialized changed. The "broken" version is https://github.com/gabe-l-hart/llama.cpp/blob/broken/GraniteFour/src/llama-graph.cpp (based off of `ed52f36` on `master`) while the "working" version is https://github.com/gabe-l-hart/llama.cpp/blob/GraniteFour/src/llama-graph.cpp (based off of `f470bc3` on `master`).

ggerganov commented on src/llama-kv-cache-hybrid-recurrent.cpp:
```diff
@@ -0,0 +1,252 @@
+    return true;
+}
+
+bool llama_kv_cache_hybrid_recurrent_state::apply() {
+    assert(status == LLAMA_MEMORY_STATUS_SUCCESS);
+
+    kv->get_kv_attn()     ->apply_ubatch(heads_attn[i_next], ubatches[i_next]);
+    kv->get_kv_recurrent()->find_slot(ubatches[i_next]);
+
+    return true;
+}
```
That's correct. At least the `llama_kv_cache_unified_state` state has to be initialized in a similar way as it is done in `llama_kv_cache_unified_iswa_state`:

https://github.com/ggml-org/llama.cpp/blob/fb85a288d72abbd5e5daa8de96e6f8bfa7b5ab46/src/llama-kv-cache-unified-iswa.cpp#L216-L232

gabe-l-hart replied:
That was it!! Thank you thank you! Will update this PR shortly

gabe-l-hart committed:
fix: Fix resize vs reserve and skip null tensors in size computation

https://github.com/ggml-org/llama.cpp/pull/13979/files#r2149469788

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>
Co-Authored-By: @younesbelkada

gabe-l-hart committed:
fix: Fix initialization of child states

Since initially writing this PR, the logic in the child state types changed
such that using the "init full" signature and keeping the ubatches on the
parent struct no longer worked.

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
refactor: Use a common build_recurrent_state method that is cache-agnostic

This reduces the code duplication between the different build_rs impls and
also retains a similar signature to the previous build_recurrent_state
method while standardizing on the input-dispatched build_rs implementation.

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov committed:
recurrent : rework graph inputs + add TODOs

ggml-ci

compilade commented:
> For example `n_kv` for the recurrent state does not make sense - there are no keys and values in it.

I suggest `n_rs`, which was used in #7531 exactly like `n_kv` is currently used for the recurrent cache on `master`.

gabe-l-hart committed:
Merge pull request #2 from ggml-org/gabe-l-hart/HybridRecurrentCache

recurrent : rework graph inputs + add TODOs

gabe-l-hart committed:
refactor: Make status and child states const in hybrid and iswa

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
In the rename, one thing this would open up is renaming `llama_kv_cache_unified` -> `llama_memory_kv_cache` and `llama_kv_cache_unified_iswa` -> `llama_memory_kv_cache_iswa` for consistency with other `llama_memory_*` classes. Thoughts?

gabe-l-hart committed:
refactor: Rename llama_kv_cache_[recurrent|hybrid_recurrent] to remove kv cache

This removes the notion of "kv" from the interface names for these memory
types. There are still many references to kv in the implementation of the
recurrent memory which will need further adjustment.

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

compilade commented:
> Another one as I'm working through it: in `hparams` we've got `n_embd_[k|v]_s` which I _think_ would become `n_embd_[conv|ssm]`?

Or maybe `n_embd_[r|s]` (since RWKV doesn't really call it's states `conv` or `ssm`, but does use these functions)

gabe-l-hart commented:
>Or maybe n_embd_[r|s] (since RWKV doesn't really call its states conv or ssm, but does use these functions)

Ah, interesting. I was basing off of how the mamba models use `get_[k|v]_l` which is probably not sufficiently robust.

gabe-l-hart committed:
refactor!: Rename all k/v related values for recurrent/hybrid to r/s

Anywhere that "kv_<state|cell|size|etc>" is used, I've used the more
generic "mem_" prefix. The specifics of "k" (key) translate to "r"
(recurrent state) and "v" (value) translate to "s" (state-space embedding
states).

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
Ok, renaming complete! I think it's a lot cleaner now, but would definitely love the extra eyes on anywhere I might have missed.

ggerganov commented on src/llama-hparams.h:
```diff
@@ -181,10 +184,13 @@ struct llama_hparams {
+    uint32_t n_embd_s() const;
+
+    // whether or not the given layer is recurrent (for hybrid models)
+    bool recurrent_layer(uint32_t il) const;
```
```suggestion
    bool is_recurrent(uint32_t il) const;
```

ggerganov commented on src/llama-arch.h:
```diff
@@ -439,3 +440,6 @@ const char * llm_arch_name(llm_arch arch);
 
 const llm_tensor_info & llm_tensor_info_for(llm_tensor tensor);
+
+bool llm_arch_is_recurrent(const llm_arch& arch);
+bool llm_arch_is_hybrid   (const llm_arch& arch);
```
```suggestion
bool llm_arch_is_recurrent(const llm_arch & arch);
bool llm_arch_is_hybrid   (const llm_arch & arch);
```

gabe-l-hart replied:
Fixed

ggerganov commented on src/llama-memory-hybrid.cpp:
```diff
@@ -0,0 +1,247 @@
+}
+
+void llama_memory_hybrid::seq_add(llama_seq_id seq_id, llama_pos p0, llama_pos p1, llama_pos shift) {
+    mem_attn->seq_add(seq_id, p0, p1, shift);
+    mem_recurrent->seq_add(seq_id, p0, p1, shift);
```
```suggestion
    mem_attn     ->seq_add(seq_id, p0, p1, shift);
    mem_recurrent->seq_add(seq_id, p0, p1, shift);
```

ggerganov commented on src/llama-memory-hybrid.cpp:
```diff
@@ -0,0 +1,247 @@
+}
+
+void llama_memory_hybrid::state_read(llama_io_read_i & io, llama_seq_id seq_id) {
+    mem_attn     ->state_read(io, seq_id);
+    mem_recurrent->state_read(io, seq_id);
```
nit: We could change the "recurrent" suffix in this class to "recr" to make this more compact:

```suggestion
    mem_attn->state_read(io, seq_id);
    mem_recr->state_read(io, seq_id);
```

gabe-l-hart replied:
There were a couple of other spots that needed to get touched, so I did the change locally (plus I'm _guessing_ you're not too worried about your commit stats on this project on this order of magnitude)

ggerganov reviewed (approved):
This is good to merge. Thanks for the patience and catching up with all the changes on `master` 👍

gabe-l-hart commented:
> This is good to merge. Thanks for the patience and catching up with all the changes on master 👍

Thank you! I learned a _ton_ along the way. I'll make the suggested change from `recurrent` -> `recr` and kick CI one more time.

gabe-l-hart committed:
refacor: _recurrent -> _recr for brevity

It just _happens_ to have the same number of letters as _attn!

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
style: Fix spacing for ref

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
refactor: recurrent_layer() -> is_recurrent()

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
Ok, renames/refactors done. 🤞 on Ci!

ggerganov commented on src/llama-memory-recurrent.h:
```diff
@@ -109,8 +115,8 @@ class llama_kv_cache_recurrent : public llama_memory_i {
-    size_t size_k_bytes() const;
-    size_t size_v_bytes() const;
+    size_t size_r_bytes() const;
+    size_t size_s_bytes () const;
```
```suggestion
    size_t size_s_bytes() const;
```

gabe-l-hart committed:
style: Fix spacing for size_s_bytes declaration

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

compilade commented on src/llama-graph.h:
```diff
@@ -589,22 +618,62 @@ struct llm_graph_context {
-                 int32_t   n_seqs,
-                    bool   avoid_copies = false) const;
+    // TODO: avoid notion of "kv"
+    // TODO: move this implementation to llama_memory_recurrent.
+    //       this is analogous to llama_kv_cache_unified::cpy_k / cpy_v
+    //       when moving, avoid passing `ggml_cgraph` - only pass `ggml_context`. would likely need to split the
+    //         implementation in 2 separate methods. the goal is to avoid calling `ggml_build_forward_expand` in
+    //         `llama_memory_recurrent`
```
> the goal is to avoid calling `ggml_build_forward_expand`

I'm not sure that will be convenient; `build_rs` when used with `avoid_copies` pretty much *has to* have a way to order the writes.

Splitting the function in two would require handling `states_extra` at call sites, while for now it's transparently handled within `build_rs`.

ggerganov replied:
I see. It's ok to pass the `ggml_cgraph` as argument in that case.
